src/Engine/ProtoCore/DSASM/Heap.cs

```diff
@@ -416,7 +421,29 @@ public StackValue AllocatePointer(int size, MetaData metadata)
         private StackValue AllocateStringInternal(string str, bool isConstant)
         {
             int index;
-            if (!stringTable.TryGetPointer(str, out index))
+            if (stringTable.TryGetPointer(str, out index)) {
+                // Any existing heap elements, marked as white, that are in the sweepSet and that are referenced during the sweep cycle will be marked as Black.
+                // This will ensure that no reachable data is mistakenly cleaned up.
+                bool isDuringGCPropagateCycle = gcState == GCState.Propagate;
+                bool isValidHeapIndex = index >= 0 && index < heapElements.Count;
+                if (isDuringGCPropagateCycle && isValidHeapIndex)
+                {
+                    var he = heapElements[index];
```

Comment:
Does the index found in the `stringTable` match that in the list of `heapElements`? Where do we ensure that the index in the `stringTable` points to the same element at the same index in the `heapElements` list?

Reply:
The index in the stringTable and the heap are kept in sync by always cleaning both at the same time.
When a heap element is cleaned up, the stringTable also cleans the matching data.